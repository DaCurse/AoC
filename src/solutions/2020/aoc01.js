export function partOne(inputs) {
  // We will store numbers in a set until we find our solution
  const prev = new Set();
  const target = 2020;

  for (let i = 0; i < inputs.length; ++i) {
    const input = inputs[i];

    // If a number exists in the set who is the difference between the target
    // and the current number, we found our pair
    if (!prev.has(target - input)) {
      prev.add(input);
    } else {
      return {
        // Returning the pair alongisde the result for propery testing
        fluff: [target - input, input],
        result: (target - input) * input,
      };
    }
  }
}
