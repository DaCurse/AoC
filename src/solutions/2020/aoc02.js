const pwRegex = /(\d+)-(\d+)\s(\w):\s(\w+)/;

export function partOne(inputs) {
  return inputs.reduce((acc, input) => {
    const [, min, max, letter, pw] = input.match(pwRegex);
    const count = pw.split(letter).length - 1;

    return count >= Number(min) && count <= Number(max) ? acc + 1 : acc;
  }, 0);
}
