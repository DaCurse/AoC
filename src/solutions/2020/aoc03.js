export function partOne(inputs) {
  const lineLength = inputs[0].length;
  let x = 0;
  let y = 0;
  let trees = 0;
  while (y < inputs.length - 1) {
    x = (x + 3) % lineLength;
    if (inputs[++y][x] === '#') {
      trees++;
    }
  }

  return trees;
}
