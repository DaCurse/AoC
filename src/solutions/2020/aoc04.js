export function parseFields(raw) {
  return raw.split('\n\n').map((p) => p.split(/\s|:/));
}

export function partOne(inputs) {
  const requiredFields = ['byr', 'iyr', 'eyr', 'hgt', 'hcl', 'ecl', 'pid'];
  return inputs.reduce(
    (acc, passport) =>
      requiredFields.every((field) => passport.includes(field)) ? acc + 1 : acc,
    0
  );
}
