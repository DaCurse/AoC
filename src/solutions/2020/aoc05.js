const map = {
  B: '1',
  R: '1',
  F: '0',
  L: '0',
};

function calcIDs(inputs) {
  return inputs.map((input) =>
    parseInt(
      input.replace(/[BR]|[FL]/g, (char) => map[char]),
      2
    )
  );
}

export function partOne(inputs) {
  return Math.max(...calcIDs(inputs));
}

export function partTwo(inputs) {
  const xor = (a, b) => a ^ b;
  const ids = calcIDs(inputs);
  const max = Math.max(...ids);
  const min = Math.min(...ids);

  return (
    ids.reduce(xor) ^
    Array.from({ length: max }, (_, i) => i + 1).reduce(xor) ^
    Array.from({ length: min - 1 }, (_, i) => i + 1).reduce(xor)
  );
}
