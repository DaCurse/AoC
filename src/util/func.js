/**
 * Returns the sum of `n1` and `n2`
 * @param {number} n1
 * @param {number} n2
 */
export function sum(n1, n2) {
  return n1 + n2;
}

/**
 * Returns the product of `n1` and `n2`
 * @param {number} n1
 * @param {number} n2
 */
export function product(n1, n2) {
  return n1 * n2;
}

/**
 * Returns the difference between `n1` and `n2`
 * @param {number} n1
 * @param {number} n2
 */
export function diff(n1, n2) {
  return n1 - n2;
}

/**
 * Returns the quotient from dividing `n1` by `n2`
 * @param {number} n1
 * @param {number} n2
 */
export function quotient(n1, n2) {
  return n1 / n2;
}
