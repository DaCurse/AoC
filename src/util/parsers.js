/**
 * Parses a list of strings to an array with a newline as the delimiter, getting
 * rid of empty results
 * @param {string} raw List of strings
 */
export function parseListOfStrings(raw) {
  return raw.split(/\r?\n/).filter((s) => s);
}

/**
 * Parses a list of groups seperated each by a blank line
 * @param {string} raw List of groups
 */
export function parseListOfGroups(raw) {
  return raw.split(/(?:\r?\n){2}/).filter((s) => s);
}

/**
 * Parses list of strings to an array of integers
 * @param {string} raw List of strings
 */
export function parseListOfInts(raw) {
  return parseListOfStrings(raw).map((n) => parseInt(n, 10));
}
