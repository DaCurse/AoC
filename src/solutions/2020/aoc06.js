import { parseListOfStrings } from '../../util/parsers';

export function partOne(inputs) {
  return inputs.reduce(
    (total, input) => total + new Set(input.replace(/\s/g, '')).size,
    0
  );
}

export function partTwo(inputs) {
  return inputs.reduce((total, input) => {
    const unanimous = parseListOfStrings(input)
      .map((s) => Array.from(s))
      .reduce((intersection, answers) =>
        intersection.filter((a) => answers.includes(a))
      );
    return total + unanimous.length;
  }, 0);
}
