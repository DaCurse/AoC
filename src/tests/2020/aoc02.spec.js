import { partOne } from '../../solutions/2020/aoc02';
import getInputs from '../../util/inputs';
import { parseListOfStrings } from '../../util/parsers';

describe('Day 2 solution', () => {
  test('Part one', async () => {
    const inputs = await getInputs(2, parseListOfStrings);
    const result = partOne(inputs);
    console.log('Day 2 Part 1 solution:', result);

    expect(result).toMatchSnapshot();
  });
});
