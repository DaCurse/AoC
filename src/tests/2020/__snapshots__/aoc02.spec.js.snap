// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Day 2 solution Part one 1`] = `591`;
